### ISSUE TITLE
Error while using hello-engine.js

### ISSUE THREAD

**Issue by L3x1k0n** [qlik-oss/core-get-started#169]
### Description

Get an Error while the "Getting Started" guide. At the moment i load "hello-engine"

### Steps to Reproduce

Just followed the Guide on the Qlikcore Getting Started Website
https://qlikcore.com/tutorials/hello-engine/

##### Expected behavior

The output from the "hello-engine" guide 
```
npm run hello-engine
Creating and opening session.
Engine version retrieved: 12.91.0
Session closed.
```

##### Actual behavior

```
> node ./src/hello-engine/hello-engine.js

Creating and opening session.
Whoops! An error occurred. ErrorEvent {
  target:
   WebSocket {
     _events:
      { open: [Function],
        close: [Function],
        error: [Function],
        message: [Function] },
     _eventsCount: 4,
     _maxListeners: undefined,
     readyState: 3,
     protocol: '',
     _binaryType: 'nodebuffer',
     _closeFrameReceived: false,
     _closeFrameSent: false,
     _closeMessage: '',
     _closeTimer: null,
     _closeCode: 1006,
     _extensions: {},
     _isServer: false,
     _receiver: null,
     _sender: null,
     _socket: null,
     url: 'ws://localhost:19076/app',
     _req: null },
  type: 'error',
  message: 'connect ECONNREFUSED 127.0.0.1:19076',
  error:
   { Error: connect ECONNREFUSED 127.0.0.1:19076
       at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1113:14)
     errno: 'ECONNREFUSED',
     code: 'ECONNREFUSED',
     syscall: 'connect',
     address: '127.0.0.1',
     port: 19076 } }
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! core-get-started@0.0.1 hello-engine: `node ./src/hello-engine/hello-engine.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the core-get-started@0.0.1 hello-engine script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

```

##### Library
```
[x ] Node.js
[ ] Browser
```
##### Operating system
```
[x ] Windows
[ ] OSX
[ ] Linux
```
### Versions

* Node.js: 
* Browser: Google Chrome  68.0.3440.106 
* Operating system: Windows 10

Maybe someone can help me. Im always getting this error in the Getting Started guide from Qlik Core. The Server is started in Docker and then i try to run the "hello-engine.js" but always get the Error Message above.

**Comment by FredrikFolkesson** [comment]
Can you show the output if you run `docker ps`?

**Comment by L3x1k0n** [comment]
Docker says this :
```
CONTAINER ID        IMAGE                      COMMAND                  CREATED             STATUS              PORTS                NAMES
ba72445f00fa        qlikcore/engine:12.225.0   "/usr/local/bin/engi…"   32 minutes ago      Up 32 minutes       9076/tcp, 9090/tcp   wizardly_swirles
```

**Comment by FredrikFolkesson** [comment]
It does not look like you are forwarding from port `19076` on your machine to port `9076` on the engine.

Did you run the command `ACCEPT_EULA=yes docker-compose up -d` ?

**Comment by FredrikFolkesson** [comment]
`docker run` is used when you are _not_ using a compose file (usually when you just want to run one container. While docker-compose files are usually used when you want to couple a few containers together so they can talk over the same network (or to save some settings as we have done here.

So you could start the engine by running the docker-compose with:
`ACCEPT_EULA=yes docker-compose up -d`

OR

you could run:
`docker run -p 19076:9076 qlikcore/engine:12.225.0 -S AcceptEULA=yes` (on this you would later have to add the volumes, something we have already done in the docker-compose file.

So my reccomendation would be that you stop your running containres and then start them using the compose with the command `ACCEPT_EULA=yes docker-compose up -d`

**Comment by sublibra** [comment]
Closing this due to inactivity. Please reopen or contact us if you still have problems running it.

**Comment by vivek-gajjar** [comment]
Hi, I'm hoping you can help. I'm having the same issue as well and I've run the ACCEPT_EULA=yes docker-compose up -d command and I've also run the docker ps command and I have the following: 
`PS C:\Users\vivek.gajjar\documents\qlik\core\core-get-started> docker ps
CONTAINER ID        IMAGE                      COMMAND                  CREATED             STATUS              PORTS
                            NAMES
4bae1bcf2b97        qlikcore/engine:12.401.0   "/usr/local/bin/engi"   2 minutes ago       Up 2 minutes        9090/tcp,
 0.0.0.0:19076->9076/tcp   coregetstarted_engine_1`

I've also run the npm install but I'm getting the same error, any suggestions?

**Comment by sublibra** [comment]
Hello Vivek,

**Comment by vivek-gajjar** [comment]
Hi Jonas

Thanks for your prompt response. I had just composed down however have re-run it and changed the container Id 

Please see the logs below:
`{"timestamp":"2019-07-03T12:33:49.684Z","log_type":"System","logseverity":"INFO"
,"thread_id":3579240000,"user_id":"Internal\\Engine","tenant_id":"default","app_
id":"","message":"Engine started with command line: --MigrationPort -1 -S QvLogU
seStdoutLogger=1 -S QvLogHandleRequestOnExit=1 -S AllowExternalSocket=1 -S Enabl
eSwagger=1 -S EnablePrometheus=1 -S PrometheusServicePort=9090 -S PerformanceLog
Verbosity=0 -S UseCGroupsStats=1 -S AdjustMemUseFromOs=0 -S EnableCrashDump=1 -S
 EnableScriptLogging=2 -S AcceptEULA=yes "}
{"timestamp":"2019-07-03T12:33:49.686Z","log_type":"Start","logseverity":"INFO",
"thread_id":3579240000,"message":"Invalid parameter handler set"}
{"timestamp":"2019-07-03T12:33:49.687Z","log_type":"Start","logseverity":"INFO",
"thread_id":3579240000,"message":"ComputerName is 29735d8d1834"}
{"timestamp":"2019-07-03T12:33:49.687Z","log_type":"Start","logseverity":"INFO",
"thread_id":3579240000,"message":"Engine Version is 12.401.0"}
{"timestamp":"2019-07-03T12:33:49.965Z","log_type":"Start","logseverity":"INFO",
"thread_id":3579240000,"message":"Locale(s) set"}
{"timestamp":"2019-07-03T12:33:49.966Z","log_type":"Start","logseverity":"INFO",
"thread_id":3579240000,"message":"QHeap initialized"}
{"timestamp":"2019-07-03T12:33:49.967Z","log_type":"Start","logseverity":"INFO",
"thread_id":3579240000,"message":"Running in desktop mode"}
{"timestamp":"2019-07-03T12:33:49.973Z","log_type":"Start","logseverity":"INFO",
"thread_id":3579240000,"message":"Initiating server license."}
{"timestamp":"2019-07-03T12:33:49.974Z","log_type":"System","logseverity":"INFO"
,"thread_id":3579240000,"user_id":"Internal\\Engine","tenant_id":"default","app_
id":"","message":"GlobalContext: Initiating Broker default port 4848"}
{"timestamp":"2019-07-03T12:33:49.974Z","log_type":"System","logseverity":"INFO"
,"thread_id":3579240000,"user_id":"Internal\\Engine","tenant_id":"default","app_
id":"","message":"GlobalContext: Initiating DataPrepPort default port 9072"}
{"timestamp":"2019-07-03T12:33:49.977Z","log_type":"System","logseverity":"INFO"
,"thread_id":3579240000,"user_id":"Internal\\Engine","tenant_id":"default","app_
id":"","message":"WorkingSet: Configured Working Set Size is 0.676-0.870 GB (set
ting 70%-90%)"}
{"timestamp":"2019-07-03T12:33:49.980Z","log_type":"Start","logseverity":"INFO",
"thread_id":3579240000,"message":"Initiating ExtEngine."}
{"timestamp":"2019-07-03T12:33:49.982Z","log_type":"Start","logseverity":"INFO",
"thread_id":3579240000,"message":"Initialization complete."}
{"timestamp":"2019-07-03T12:33:49.983Z","log_type":"System","logseverity":"INFO"
,"thread_id":3579240000,"user_id":"Internal\\Engine","tenant_id":"default","app_
id":"","message":"QvSocket: Starting HTTP request handler - 0x7fc4d32a8620"}
{"timestamp":"2019-07-03T12:33:49.984Z","log_type":"System","logseverity":"INFO"
,"thread_id":3579240000,"user_id":"Internal\\Engine","tenant_id":"default","app_
id":"","message":"Port Listen: IPv4 bind successful on Port 9076"}
{"timestamp":"2019-07-03T12:33:49.985Z","log_type":"System","logseverity":"INFO"
,"thread_id":3579240000,"user_id":"Internal\\Engine","tenant_id":"default","app_
id":"","message":"Port Listen: IPv6 bind successful on Port 9076"}
{"timestamp":"2019-07-03T12:33:49.986Z","log_type":"System","logseverity":"INFO"
,"thread_id":3579240000,"user_id":"Internal\\Engine","tenant_id":"default","app_
id":"","message":"System: Starting - Purge: Starting"}
{"timestamp":"2019-07-03T12:33:49.987Z","log_type":"System","logseverity":"INFO"
,"thread_id":3579240000,"user_id":"Internal\\Engine","tenant_id":"default","app_
id":"","message":"System: Starting - ServerWatchDog: Starting"}
{"timestamp":"2019-07-03T12:33:50.991Z","log_type":"System","logseverity":"INFO"
,"thread_id":3509102336,"user_id":"Internal\\Engine","tenant_id":"default","app_
id":"","message":"Mount: Found Mount  at location /home/engine/Qlik/Sense/Apps
browsable"}`

Thanks in advance.

**Comment by sublibra** [comment]
Looks good so far. Did you run `npm run hello-engine` after the engine restart?

**Comment by vivek-gajjar** [comment]
Yes I did unfortunately the same error:
`$ npm run hello-engine

> core-get-started@0.0.1 hello-engine C:\users\vivek.gajjar\Documents\Qlik\Core\
core-get-started
> node ./src/hello-engine/hello-engine.js

Creating and opening session.
Whoops! An error occurred. ErrorEvent {
  target:
   WebSocket {
     _events:
      [Object: null prototype] {
        open: [Function],
        close: [Function],
        error: [Function],
        message: [Function] },
     _eventsCount: 4,
     _maxListeners: undefined,
     readyState: 3,
     protocol: '',
     _binaryType: 'nodebuffer',
     _closeFrameReceived: false,
     _closeFrameSent: false,
     _closeMessage: '',
     _closeTimer: null,
     _closeCode: 1006,
     _extensions: {},
     _receiver: null,
     _sender: null,
     _socket: null,
     _bufferedAmount: 0,
     _isServer: false,
     _redirects: 0,
     url: 'ws://localhost:19076/app',
     _req: null },
  type: 'error',
  message: 'connect ECONNREFUSED 127.0.0.1:19076',
  error:
   { Error: connect ECONNREFUSED 127.0.0.1:19076
       at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1106:14)
     errno: 'ECONNREFUSED',
     code: 'ECONNREFUSED',
     syscall: 'connect',
     address: '127.0.0.1',
     port: 19076 } }
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! core-get-started@0.0.1 hello-engine: `node ./src/hello-engine/hello-eng
ine.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the core-get-started@0.0.1 hello-engine script.
npm ERR! This is probably not a problem with npm. There is likely additional log
ging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\vivek.gajjar\AppData\Roaming\npm-cache\_logs\2019-07-03T12
_40_33_573Z-debug.log`

I've also check the log file as suggested of which the content is:
`0 info it worked if it ends with ok
1 verbose cli [ 'C:\\Program Files\\nodejs\\node.exe',
1 verbose cli   'C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js',
1 verbose cli   'run',
1 verbose cli   'hello-engine' ]
2 info using npm@6.9.0
3 info using node@v10.16.0
4 verbose run-script [ 'prehello-engine', 'hello-engine', 'posthello-engine' ]
5 info lifecycle core-get-started@0.0.1~prehello-engine: core-get-started@0.0.1
6 info lifecycle core-get-started@0.0.1~hello-engine: core-get-started@0.0.1
7 verbose lifecycle core-get-started@0.0.1~hello-engine: unsafe-perm in lifecycle true
8 verbose lifecycle core-get-started@0.0.1~hello-engine: PATH: C:\Program Files\nodejs\node_modules\npm\node_modules\npm-lifecycle\node-gyp-bin;C:\users\vivek.gajjar\Documents\Qlik\Core\core-get-started\node_modules\.bin;C:\Users\vivek.gajjar\bin;C:\Program Files\Git\mingw64\bin;C:\Program Files\Git\usr\local\bin;C:\Program Files\Git\usr\bin;C:\Program Files\Git\usr\bin;C:\Program Files\Docker Toolbox;C:\Users\vivek.gajjar\bin;C:\Program Files\Git\mingw64\bin;C:\Program Files\Git\usr\local\bin;C:\Program Files\Git\usr\bin;C:\Program Files\Git\usr\bin;C:\Program Files\Git\mingw64\bin;C:\Program Files\Git\usr\bin;C:\Users\vivek.gajjar\bin;C:\Program Files (x86)\Common Files\Oracle\Java\javapath;C:\ProgramData\Oracle\Java\javapath;C:\windows\system32;C:\windows;C:\windows\System32\Wbem;C:\windows\System32\WindowsPowerShell\v1.0;C:\Program Files\Intel\WiFi\bin;C:\Program Files\Common Files\Intel\WirelessCommon;C:\Program Files (x86)\ATI Technologies\ATI.ACE\Core-Static;E:\Tools\Alteryx\bin;C:\Program Files (x86)\Microsoft SQL Server\Client SDK\ODBC\130\Tools\Binn;C:\Program Files (x86)\Microsoft SQL Server\130\Tools\Binn;C:\Program Files (x86)\Microsoft SQL Server\130\DTS\Binn;C:\Program Files (x86)\Microsoft SQL Server\130\Tools\Binn\ManagementStudio;C:\Program Files\Snowflake SnowSQL;C:\Program Files\PuTTY;C:\Program Files\dotnet;C:\Program Files\nodejs;C:\Program Files\Git\cmd;C:\Users\vivek.gajjar\AppData\Local\Programs\Microsoft VS Code\bin;C:\Users\vivek.gajjar\AppData\Local\Programs\Python\Python37-32;C:\Users\vivek.gajjar\AppData\Roaming\npm;C:\Program Files\Docker Toolbox;C:\Program Files\Git\usr\bin\vendor_perl;C:\Program Files\Git\usr\bin\core_perl;C:\Program Files\Git\usr\bin\vendor_perl;C:\Program Files\Git\usr\bin\core_perl
9 verbose lifecycle core-get-started@0.0.1~hello-engine: CWD: C:\users\vivek.gajjar\Documents\Qlik\Core\core-get-started
10 silly lifecycle core-get-started@0.0.1~hello-engine: Args: [ '/d /s /c', 'node ./src/hello-engine/hello-engine.js' ]
11 silly lifecycle core-get-started@0.0.1~hello-engine: Returned: code: 1  signal: null
12 info lifecycle core-get-started@0.0.1~hello-engine: Failed to exec hello-engine script
13 verbose stack Error: core-get-started@0.0.1 hello-engine: `node ./src/hello-engine/hello-engine.js`
13 verbose stack Exit status 1
13 verbose stack     at EventEmitter.<anonymous> (C:\Program Files\nodejs\node_modules\npm\node_modules\npm-lifecycle\index.js:301:16)
13 verbose stack     at EventEmitter.emit (events.js:198:13)
13 verbose stack     at ChildProcess.<anonymous> (C:\Program Files\nodejs\node_modules\npm\node_modules\npm-lifecycle\lib\spawn.js:55:14)
13 verbose stack     at ChildProcess.emit (events.js:198:13)
13 verbose stack     at maybeClose (internal/child_process.js:982:16)
13 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:259:5)
14 verbose pkgid core-get-started@0.0.1
15 verbose cwd C:\users\vivek.gajjar\Documents\Qlik\Core\core-get-started
16 verbose Windows_NT 6.1.7601
17 verbose argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "run" "hello-engine"
18 verbose node v10.16.0
19 verbose npm  v6.9.0
20 error code ELIFECYCLE
21 error errno 1
22 error core-get-started@0.0.1 hello-engine: `node ./src/hello-engine/hello-engine.js`
22 error Exit status 1
23 error Failed at the core-get-started@0.0.1 hello-engine script.
23 error This is probably not a problem with npm. There is likely additional logging output above.
24 verbose exit [ 1, true ]`

**Comment by axelssonHakan** [comment]
The **ECONNREFUSED** (=Connection refused) in the error message indicates that there is something blocking the connection.
Is there a firewall running on the computer blocking?

**Comment by vivek-gajjar** [comment]
I've just checked the firewall on my computer and it seems to be switched off

**Comment by axelssonHakan** [comment]
You could try connecting with http to the QIX Engine:
- http://localhost:9076/health -> Returns 200/OK
- http://localhost:9076/openAPI -> Returns the openAPI spec

If you aren't able to connect to the engine it's probably blocked by a firewall or there is a web-proxy defined in your network setting. If a web-proxy is used you need to add a "rule" to ignore proxying then going to `localhost`

**Comment by vivek-gajjar** [comment]
Ok so I think I've found the issue. In the hello-engine.js file the issue was with the port number in the  `url: 'ws://localhost:19076/app',` bit when I changed it to `url: 'ws://localhost:9076/app',` so by changing the port number from 19076 to 9076 it's managed to work. 

`$ npm run hello-engine

> core-get-started@0.0.1 hello-engine C:\users\vivek.gajjar\Documents\Qlik\Core\
core-get-started
> node ./src/hello-engine/hello-engine.js

Creating and opening session.
Engine version retrieved: 12.287.2
Session closed.`

However I've got another issue now in that when I'm trying the hello-data example I'm getting an error and I think it's because the data connections cannot be made.

`PS C:\Users\vivek.gajjar\documents\qlik\core\core-get-started> npm run hello-data

> core-get-started@0.0.1 hello-data C:\Users\vivek.gajjar\documents\qlik\core\core-get-started
> node ./src/hello-data/hello-data.js

Creating session app on engine.
Engine version retrieved: 12.287.2
Creating data connection to local files.
Whoops! An error occurred. { Error: Invalid path
    at Intercept.errorInterceptor (C:\Users\vivek.gajjar\documents\qlik\core\core-get-started\node_modules\enigma.js\eni
gma.js:2549:19)
    at process._tickCallback (internal/process/next_tick.js:68:7) code: 4, parameter: 'Invalid path' }
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! core-get-started@0.0.1 hello-data: `node ./src/hello-data/hello-data.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the core-get-started@0.0.1 hello-data script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\vivek.gajjar\AppData\Roaming\npm-cache\_logs\2019-07-04T11_10_52_160Z-debug.log`

**Comment by hrigner** [comment]
Are you using the docker-compose.yml found in the root in this repo? Seems that you are not exposing the right port (19076) and also not mounting the data volume needed for hello-data to work as expected.

**Comment by hrigner** [comment]
And also, "Engine version retrieved: 12.287.2" says that it's not the same engine as the output from `docker ps` command that you posted yesterday. Do you have multiple engines running?

**Comment by joitlik** [comment]
Hello everyone! i had the same error and then just run 'ACCEPT_EULA=yes docker-compose up -d' then i  ran  'npm run hello-engine' command again and Wuala, now it's working!

**Comment by sublibra** [comment]
Closing this due to inactivity. Please open again if you haven't solved the issue.
